jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +190,370 @@ default void clear()
+                    for (RetainableByteBuffer buffer : _buffers)
+                    {
+                        buffer.putTo(byteBuffer);
+                        buffer.clear();
+                        buffer.release();
+                    }
+                    BufferUtil.flipToFlush(byteBuffer, 0);
+                    _buffers.clear();
+                    _buffers.add(combinedBuffer);
+                    _canAggregate = true;
+                    yield combinedBuffer.getByteBuffer();
+                }
+            };
+        }
+
+        /**
+         * {@inheritDoc}
+         * @throws ArithmeticException if the length of this {@code Accumulator} is greater than {@link Integer#MAX_VALUE}
+         */
+        @Override
+        public int remaining()
+        {
+            return Math.toIntExact(remainingLong());
+        }
+
+        public long remainingLong()
+        {
+            long length = 0;
+            for (RetainableByteBuffer buffer : _buffers)
+                length += buffer.remaining();
+            return length;
+        }
+
+        /**
+         * {@inheritDoc}
+         * @throws ArithmeticException if the maxLength of this {@code Accumulator} is greater than {@link Integer#MAX_VALUE}.
+         */
+        @Override
+        public int capacity()
+        {
+            return Math.toIntExact(capacityLong());
+        }
+
+        public long capacityLong()
+        {
+            return _maxLength;
+        }
+
+        @Override
+        public boolean release()
+        {
+            clear();
+            return true;
+        }
+
+        @Override
+        public void clear()
+        {
+            for (RetainableByteBuffer buffer : _buffers)
+                buffer.release();
+            _canAggregate = false;
+            _buffers.clear();
+        }
+
+        @Override
+        public boolean append(ByteBuffer bytes)
+        {
+            int remaining = bytes.remaining();
+            int length = ensureMaxLength(remaining);
+            // if the length was restricted by maxLength, slice the bytes smaller
+            ByteBuffer out = bytes;
+            if (length < remaining)
+            {
+                ByteBuffer slice = bytes.slice();
+                slice.limit(slice.position() + length);
+                bytes.position(bytes.position() + length);
+                out = slice;
+            }
+            RetainableByteBuffer buffer = _pool.acquire(length, _direct);
+            buffer.append(out);
+            _buffers.add(buffer);
+            _canAggregate = true;
+            return !bytes.hasRemaining();
+        }
+
+        @Override
+        public boolean append(RetainableByteBuffer bytes)
+        {
+            int remaining = bytes.remaining();
+            int length = ensureMaxLength(remaining);
+
+            // If we cannot retain, then try aggregation into the last buffer
+            if (!bytes.canRetain() && _canAggregate)
```

Comment:
I'm not sure I understand that `canRetain()` / `canAggregate` logic. If `canRetain()` returns false, this means that RBB's lifecycle is left to the GC, so I don't get why we should aggregate in that case.

Reply:
So you are saying that we can assume a non-retainable buffer can be kept by reference and not released, but just GC'd.

Hmmm maybe.    I guess we should say that in the javadoc for canRetain!